File: python/Linked List/reverse_nodes_k_groups.py
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        """
        Reverse nodes in a linked list k at a time.
        
        Strategy:
        1. Use a dummy node to handle the head of the list
        2. For each group of k nodes:
           - Find the kth node
           - Reverse the group
           - Connect the reversed group to the rest of the list
        
        Time Complexity: O(n) where n is the number of nodes
        Space Complexity: O(1) as we only use pointers
        
        Example:
        Input: 1->2->3->4->5, k=2
        Output: 2->1->4->3->5
        """
        # Create dummy node to handle the head of the list
        dummy = ListNode(0, head)
        groupPrev = dummy

        while True:
            # Get the kth node from the current position
            kth = self.getKth(groupPrev, k)  # type: ignore
            if not kth:
                break
            groupNext = kth.next  # type: ignore

            # Reverse the group
            prev, curr = kth.next, groupPrev.next  # type: ignore
            while curr != groupNext:
                tmp = curr.next  # type: ignore
                curr.next = prev  # type: ignore
                prev = curr
                curr = tmp

            # Connect the reversed group to the rest of the list
            tmp = groupPrev.next  # type: ignore
            groupPrev.next = kth  # type: ignore
            groupPrev = tmp
            
        return dummy.next

    def getKth(self, curr: ListNode, k: int) -> Optional[ListNode]:
        """
        Get the kth node from the current position.
        
        Args:
            curr: Current node
            k: Number of nodes to traverse
            
        Returns:
            The kth node or None if not enough nodes
        """
        while curr and k > 0:
            curr = curr.next  # type: ignore
            k -= 1
        return curr

# -------------------
# Helper Functions for Testing
# -------------------
def list_to_linkedlist(lst):
    dummy = ListNode()
    cur = dummy
    for val in lst:
        cur.next = ListNode(val)
        cur = cur.next
    return dummy.next

def linkedlist_to_list(node):
    result = []
    while node:
        result.append(node.val)
        node = node.next
    return result

# -------------------
# Test Cases
# -------------------
def test_reverse_k_group():
    s = Solution()
    # Test case 1: k = 2
    l = [1,2,3,4,5]
    head = list_to_linkedlist(l)
    result = s.reverseKGroup(head, 2)
    assert linkedlist_to_list(result) == [2,1,4,3,5]
    # Test case 2: k = 3
    l = [1,2,3,4,5]
    head = list_to_linkedlist(l)
    result = s.reverseKGroup(head, 3)
    assert linkedlist_to_list(result) == [3,2,1,4,5]
    # Test case 3: k = 1 (no change)
    l = [1,2,3]
    head = list_to_linkedlist(l)
    result = s.reverseKGroup(head, 1)
    assert linkedlist_to_list(result) == [1,2,3]
    # Test case 4: k equals list length
    l = [1,2,3,4]
    head = list_to_linkedlist(l)
    result = s.reverseKGroup(head, 4)
    assert linkedlist_to_list(result) == [4,3,2,1]
    # Test case 5: List length not a multiple of k
    l = [1,2,3,4,5,6,7]
    head = list_to_linkedlist(l)
    result = s.reverseKGroup(head, 3)
    assert linkedlist_to_list(result) == [3,2,1,6,5,4,7]
    # Test case 6: Empty list
    l = []
    head = list_to_linkedlist(l)
    result = s.reverseKGroup(head, 2)
    assert linkedlist_to_list(result) == []
    print("All test cases passed!")

if __name__ == "__main__":
    test_reverse_k_group()

File: C++/Linked List/reverse_nodes_k_groups.cpp
#include <bits/stdc++.h>  // includes all standard C++ headers
using namespace std;

/**
 * Definition for singly-linked list.
 */
struct ListNode {
    int val;
    ListNode *next;

    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    /**
     * Reverse nodes in k-sized groups.
     *
     * Strategy:
     * 1. Use a dummy node to simplify operations at the head.
     * 2. For each group of k nodes:
     *      - Find the kth node
     *      - Reverse the group
     *      - Connect the reversed group back to the list
     *
     * Time Complexity: O(n) — each node is visited once
     * Space Complexity: O(1) — uses only pointers
     */
    ListNode* reverseKGroup(ListNode* head, int k) {
        // dummy node to simplify edge cases
        ListNode dummy(0, head);
        ListNode* groupPrev = &dummy;

        while (true) {
            // find kth node from current position
            ListNode* kth = getKth(groupPrev, k);
            if (!kth) break;  // not enough nodes to reverse
            ListNode* groupNext = kth->next;

            // reverse current group
            ListNode* prev = groupNext;
            ListNode* curr = groupPrev->next;

            while (curr != groupNext) {
                ListNode* tmp = curr->next;
                curr->next = prev;
                prev = curr;
                curr = tmp;
            }

            // connect previous part to new head
            ListNode* tmp = groupPrev->next;
            groupPrev->next = kth;
            groupPrev = tmp;
        }

        return dummy.next;
    }

private:
    /**
     * Helper function: get the kth node from current node
     *
     * Args:
     *   curr: starting node
     *   k: number of steps to move
     *
     * Returns:
     *   pointer to kth node if exists, else nullptr
     */
    ListNode* getKth(ListNode* curr, int k) {
        while (curr && k > 0) {
            curr = curr->next;
            k--;
        }
        return curr;
    }
};


/*
-------------------- LINE-BY-LINE SYNTAX EXPLANATION --------------------

#include <bits/stdc++.h>
    ✅ Includes all standard C++ headers (like <vector>, <iostream>, etc.) at once.

using namespace std;
    ✅ Allows you to use names like vector, cout without writing std::vector.

struct ListNode {
    ✅ Defines a structure to represent a node of a singly-linked list.

int val;
    ✅ Stores the integer value of this node.

ListNode* next;
    ✅ Pointer to the next node in the list.

ListNode() : val(0), next(nullptr) {}
    ✅ Default constructor: sets val = 0 and next = nullptr.

ListNode(int x) : val(x), next(nullptr) {}
    ✅ Constructor that initializes val = x, next = nullptr.

ListNode(int x, ListNode* next) : val(x), next(next) {}
    ✅ Constructor with val and next pointer.

class Solution {
    ✅ Defines the class Solution.

public:
    ✅ Public methods accessible from outside the class.

ListNode* reverseKGroup(ListNode* head, int k)
    ✅ Function that reverses nodes in k-sized groups.

ListNode dummy(0, head);
    ✅ Dummy node with value 0, next pointing to head.

ListNode* groupPrev = &dummy;
    ✅ Pointer that tracks the node before the current group.

while (true)
    ✅ Infinite loop that breaks when fewer than k nodes remain.

ListNode* kth = getKth(groupPrev, k);
    ✅ Find the kth node from current position.

if (!kth) break;
    ✅ If kth is nullptr (not enough nodes), exit loop.

ListNode* groupNext = kth->next;
    ✅ Next node after the group.

ListNode* prev = groupNext;
ListNode* curr = groupPrev->next;
    ✅ Initialize pointers for reversal: prev is one step beyond group.

while (curr != groupNext)
    ✅ Reverse nodes in the current group.

ListNode* tmp = curr->next;
    ✅ Save next node.

curr->next = prev;
    ✅ Reverse the link of current node.

prev = curr; curr = tmp;
    ✅ Advance both pointers.

ListNode* tmp = groupPrev->next;
    ✅ Save old start of group (which becomes the end after reversal).

groupPrev->next = kth;
    ✅ Connect previous part to new head of group.

groupPrev = tmp;
    ✅ Move groupPrev forward for next group.

return dummy.next;
    ✅ Return new head of the reversed list.

private:
    ✅ Members below are accessible only within the class.

ListNode* getKth(ListNode* curr, int k)
    ✅ Finds the kth node from curr, or nullptr if not enough nodes.

while (curr && k > 0)
    ✅ Traverse k steps forward.

return curr;
    ✅ Return the kth node or nullptr.

*/
